### ISSUE TITLE
More forgiving answer checking for free text input

### ISSUE THREAD

**Issue by sibrenvanvliet** [zeeguu-ecosystem/Exercises-Platform#114]
(This should be in the Short Programming Project, but I cannot seem to add or move cards myself.)

Sometimes the learner types in an answer that a human would consider "close enough", but is rejected by the computer, because it judges very strictly. A couple of examples from my own experience:
1. I translated the German "gesonnenen" to Dutch "gezinde", but the computer wanted "gezind". They are conjugations of the same adjective, and "gezinde" actually fit bitter in the context.
2. I translated the German "gehen" to "go", but the computer wanted "to go".
3. I wrote "Nachlasse" but the computer wanted "Nachlässe".

Problem 2 is probably the most difficult to solve, because this would require the program to understand that 'to' could be omitted from an infinitive verb. If needing to write 'to' is consistent for all exercises, I think this could be considered something the user should just keep in mind.

In Problems 1 and 3, the user could get a "close enough" grade if the edit distance of the given answer and the real answer was smaller than a certain threshold value. But using edit distances is risky. It is nice for being forgiving about typos, such as writing "begining" instead of "beginning". The risk is that a word with a completely different meaning is accepted, such as "comb" instead of "bomb".

Problem 3 is easier to solve, by accepting accented letters that look like an unaccented letter to be swapped. (Even if, officially, they are different letters.) We could have a list of characters like **á**, **à**, **ä**, **ã**, **å** and accept them to be typed as **a**. We could even ignore these in the edit distance calculation.

**Comment by mircealungu** [comment]
I just sent you an invite to become collaborator to the repo. Can you add stuff to the project now?

**Comment by sibrenvanvliet** [comment]
Yes, thank you :)

**Comment by mircealungu** [comment]
Good analysis Sibren! 

Problem 3. I agree with you. I wonder if there is a js library for this kind of stuff! 

Problem 2: I think there are very few words which are in this situation. 'to' helps in forming the infinitive. and missing it or adding it even if it's not there should not penalize the learner. Besides this, in English I can only think of two other particles that could be in the same situation: the definite (the) and indefinite (a) articles. 

I think it would be the same in other languages. E.g. French: to = à, the = (le | la), a = (un, une) 

For 1, if your translation fit better in the context, then maybe we should allow the learner to say: "This is the right solution in this context" and replace the translation we have with the better translation they propose. What do you think?

**Comment by sibrenvanvliet** [comment]
There is a library on NPM for Problem 3: [remove-accents](https://www.npmjs.com/package/remove-accents). It converts accented strings to unaccented strings (for example "áàäãå" becomes "aaaaa"). During the meeting, you did mention that we wanted to include as few third party libraries as possible. Should we include this?

I searched on NPM for Problem 2 a bit, but I get the feeling that most plugins there are for English only. Because Zeeguu currently supports a limited number of languages, a possible approach is to manually make a small look-up table of 'ignorable' words in each language. Something that could look like:
1. English: "to", "the", "an", "a".
2. French: "le", "la", "un", "une", "à".
3. etc.
Because the number of words is very small for each language, I feel that this would not be too bad for scalability to more languages.

Problem 1: That is true. However, I can imagine that a user does not want to take the time to report a better solution, every time they have a one-letter difference. At the same time, we should probably avoid users abusing the system, claiming that their entry is a better solution when it actually is not. 

Some brainstorming: Maybe, if the edit distance is small, the user can get a yellow bar instead of a green bar, and the option "My translation is better" is in plain sight. But when the edit distance is large, they get a red bar, and the option "My translation is better" is only visible after clicking the drop down menu. Alternatively, we can assume for now that users are not trying to mark their wrong answers as "better translations". :)

**Comment by mircealungu** [comment]
Good point. As little third party dependencies as possible is great. At the same time, if we get to avoid defining ourselves the mapping table that they defined here: https://github.com/tyxla/remove-accents/blob/master/index.js it's probably worth it. 

Problem 2. I agree with you: defining our own solution for ignoreable-words is probably the simplest and most scalable way to go at the moment. 

Problem 1. I think we should go for a solution where we assume that the learners are not trying to cheat the system :) Also, if the edit distance is small, we probably have a typo and we don't want to provide a "my translation is better"... 

Maybe we should create three issues for these problems. The other two are straightforward and for this one we might still need to think a bit more about the best approach.

**Comment by sibrenvanvliet** [comment]
I think the most common calculation method for edit distance is the [Levenshtein Distance](https://en.wikipedia.org/wiki/Levenshtein_distance). It is also a function I have worked with before. But the [Damerau-Levenshtein Distance](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance) is an interesting alternative, because it also looks for swapping two adjacent characters, which is a common way people make typos. When given two strings, these functions simply return the edit distance between string A and string B.

If we're going to use such a function, it's important to decide on an upper limit of edit distance where the program can consider the input 'close enough'. This can either be a static number, or it can be dependent on the specific string.

Using a static number, for example 3, means that the learner always has a leeway of 3 typo'd characters per question. This is nice for small errors, for example, when different conjugations of the same adjective are confused (Dutch examples: ons/onze, rood/rode). However, if the word to translate is very short, the learner could type something completely different and still pass.

Using a string-dependent tolerance can be a solution to this drawback. For example, say it is based on the length of the solution string, and the user is allowed to make typos up to an edit distance of 20%. So if the solution string is 5 characters long, the input has an edit distance tolerance of 1. If the solution is less than 5 characters long, the learner has to get it exactly right. If the solution is 20 characters long, the learner may make typos up to an edit distance of 4.

Of course, the upper bound can potentially also be affected by other variables, such as a user setting, the learner's proficiency of a word, or the learner's proficiency of a language.

**Comment by mircealungu** [comment]
I like the Damerau-Levenshtein. Didn't know about it till now, but it's a good idea. 

What we should be aiming for in this case is precision rather than recall,
that is: better be more strict at times rather than glaringly show that we 
think 'papa' is a misspelling of 'mama'. 

intuitively i would say, that we should not even look at typos that are less
than 5 letters. there's no excuse in mistyping 'goed'. 

once we have an implementation, we can test it on past inputs that were
wrong, and see what it detects as mistypes. 

what do you think?

**Comment by sibrenvanvliet** [comment]
I agree that it's a good idea to say that learners should get short words exactly right. I realise now that the example I gave before (a learner getting striked for writing "rode" instead of "rood" even though the first would fit better in the context) is better to solve by encouraging the learner to report a solution as "bad translation". After all, the automatic translation translates words without context, so context-sensitive elements like conjugation get lost in translation.

So, we could indeed set the typo tolerance fairly low, so that we have strictness in favour of accepting blatantly wrong answers. One edit-distance per x characters, where x can be some number between say 5 or 10. Maybe this number can be dependent on the average length of the words the user is learning, their proficiency, or a user-defined setting.

**Comment by mircealungu** [comment]
One edit-distance per x characters, where x can be some number between say 5 or 10. 
I like this. Let's do it!
